File: backend/app/geo.py
"""Гео-утилиты. В проде на больших объёмах — PostGIS (ST_DWithin/индексы)."""
import math


def distance_km(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """Расстояние по формуле гаверсинуса, км."""
    r = 6371.0
    d_lat = math.radians(lat2 - lat1)
    d_lng = math.radians(lng2 - lng1)
    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(math.radians(lat1))
        * math.cos(math.radians(lat2))
        * math.sin(d_lng / 2) ** 2
    )
    return r * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

File: lib/data/models/geo.dart
import 'dart:math' as math;

/// Простая гео-точка (lat/lng). В проде — PostGIS / Яндекс.Карты.
class GeoPoint {
  const GeoPoint(this.lat, this.lng);

  final double lat;
  final double lng;

  /// Расстояние по формуле гаверсинуса, в километрах.
  double distanceKm(GeoPoint other) {
    const earthRadius = 6371.0;
    final dLat = _rad(other.lat - lat);
    final dLng = _rad(other.lng - lng);
    final a = math.sin(dLat / 2) * math.sin(dLat / 2) +
        math.cos(_rad(lat)) *
            math.cos(_rad(other.lat)) *
            math.sin(dLng / 2) *
            math.sin(dLng / 2);
    final c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a));
    return earthRadius * c;
  }

  static double _rad(double deg) => deg * (math.pi / 180.0);
}
